src/core/usubscription.rs

```diff
@@ -0,0 +1,308 @@
+/// #         ..Default::default()}).into(),
+/// #     ..Default::default()};
+/// #
+/// # let my_listener = Arc::new(MyListener::new());
+/// #
+/// # let up_client = up_client_foo::UPClientFoo::new();
+/// #
+/// use std::sync::Arc;
+/// use up_rust::{UCode, UTransport,
+///     core::usubscription::{USubscription, UnsubscribeRequest, SubscribeAttributes,
+///                           SubscriberInfo, SubscriptionRequest, SubscriptionResponse},
+/// };
+///
+/// let usub = usubscription_foo::USubscriptionFoo::new();
+///
+/// let subscriber_info = SubscriberInfo {
+///     uri: my_uuri,
+///     ..Default::default()
+/// };
+///
+/// let subscribe_attributes = SubscribeAttributes {
+///     sample_period_ms: Some(5000), // we want to hear about this every 5 seconds
+///     ..Default::default()
+/// };
+///
+/// let subscription_request = SubscriptionRequest {
+///     topic: Some(door_uuri.clone()).into(),
+///     subscriber: Some(subscriber_info.clone()).into(),
+///     attributes: Some(subscribe_attributes).into(),
+///     ..Default::default()
+/// };
+///
+/// let subscription_response = usub.subscribe(subscription_request).await;
+/// let success_code = subscription_response.status.code.enum_value_or(UCode::UNKNOWN);
+/// if success_code == UCode::OK {
+///     let register_success = up_client.register_listener(door_uuri.clone(), my_listener.clone()).await;
+/// } else {
+///     match success_code {
+///         UCode::NOT_FOUND => { /* handle topic not found */ }
+///         _ => { /* handle other error conditions */ }
+///     }
+/// }
+/// // sometime later when done with this topic
+/// let unsubscribe_request = UnsubscribeRequest {
+///     topic: Some(door_uuri.clone()).into(),
+///     subscriber: Some(subscriber_info.clone()).into(),
+///     ..Default::default()
+/// };
+/// let unsubscribe_result = usub.unsubscribe(unsubscribe_request).await;
+/// let success_code = unsubscribe_result.code.enum_value_or(UCode::UNKNOWN);
+/// if success_code == UCode::OK {
+///     let unregister_success = up_client.register_listener(door_uuri.clone(), my_listener.clone()).await;
+/// } else {
+///     // handle cases where unsubscribe failed
+/// }
+/// # }
+/// ```
+///
+/// For more information, please refer to the [uProtocol Specification](https://github.com/eclipse-uprotocol/up-spec/blob/main/up-l3/usubscription/v3/README.adoc)
+/// and [uProtocol APIs](https://github.com/eclipse-uprotocol/up-spec/blob/main/up-core-api/uprotocol/core/usubscription/v3/usubscription.proto)
+#[async_trait]
+pub trait USubscription: Send + Sync {
+    /// Subscribe to a topic, using a [`SubscriptionRequest`]
+    ///
+    /// # Parameters
+    ///
+    /// * `subscription_request` - A request to subscribe
+    ///
+    /// # Returns
+    ///
+    /// * [`SubscriptionResponse`] detailing if subscription was successful with other metadata
+    async fn subscribe(&self, subscription_request: SubscriptionRequest) -> SubscriptionResponse;
```

Comment:
```suggestion
    async fn subscribe(&self, subscription_request: SubscriptionRequest) -> Result<SubscriptionResponse, UStatus>;
```

Can we at least provide for the means to do proper response handling? Even if the service implementation would return a UStatus in the response, the local stub could still map it to the proper Result ...

Reply:
Honestly, I went back and forth on whether to handle this the same way as we did with `UTransport`'s methods.

If you look into the methods which don't return a UStatus, but return something else, e.g. a `SubscriptionResponse` it contains already a UCode as well as more specific fields as to the current `SubscriptionStatus`'s state.

Perhaps I was not modelling this correctly in my mind here.

Hmm. If we separate out the conditions, let's say like this (abbreviated pseudo code, obvs):
```
if (SubscriptionRequest is malformed || SubscriptionRequest times out) {
    return Err(UStatus::fail_with_code(...));
} else {
    // where subscription_response is the payload back from the USubscription service
    return Ok(subscription_response);
}
```
then I can definitely see it being useful to be able to model those kinds of failure conditions which do not originate from the semantics of the USubscription service itself.

I'll go ahead and update! :+1:

Reply:
> Or, to put it another way, even if we fail to subscribe, if that failure is due to the USubscription service telling us that, we'd still return Ok(). Therefore, even if someone calls subscribe() and they get back a Ok(SubscriptionResponse) they must also peer into the SubscriptionResponse's SubscriptionStatus's UCode to ensure that it's UCode::OK.

And .... we are back at https://github.com/eclipse-uprotocol/up-spec/issues/112 :-)
But I still hope, that in a not too distant future, we will see the advent of service interfaces which contain the definition of the (successful) response message only and use a dedicated (different) error reporting mechanism ...

Reply:
I hear you... it's clunky atm imho